llvm/lib/YkIR/YkIRWriter.cpp

```diff
@@ -179,14 +182,32 @@ template <class C, class E> size_t getIndex(C *Container, E *FindElement) {
   return Idx;
 }
 
-// An instruction index that uniquely identifies an Yk instruction within
+// An instruction index that uniquely identifies a Yk instruction within
 // a basic block.
 //
 // FIXME: At some point it may be worth making type-safe index types (for
 // instruction, block and function indices) and using them throughout
 // the serialiser.
 using InstIdx = size_t;
 
+// An function index that uniquely identifies a Yk function within
+// a module.
+using FuncIdx = size_t;
+
+// An basic block index that uniquely identifies a Yk basic block within
+// a function.
+using BBlockIdx = size_t;
+
+// An instruction ID that uniquely identifies a Yk instruction within a module.
+//
+// We use a raw tuple here (instead of a custom struct) so that it is hashable.
+using InstID = tuple<FuncIdx, BBlockIdx, InstIdx>;
```

Comment:
This might be a dumb question but are `InstIdx`s unique in a function or a basic block?

Reply:
An `InstIdx` is unique within a basic block. A `InstID` is unique within a module.

Reply:
Can we use `InstID` or ...? Basically 3 machine words feels like at least 1 too many. But maybe I'm wrong. And I admit that I'm prematurely optimising.

Reply:
To uniquely identify an instruction you need to know which function, block and instruction index. There's no avoiding that.

Reply:
OK.